File: 1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
int pseudoPalindromicPathsRecursive(struct TreeNode* node, int count[]) {
    if (!node) return 0;
    
    // Toggle the count for this node's value
    count[node->val] ^= 1;
    
    int ans = 0;
    // If it's a leaf node, check if the path is pseudo-palindromic
    if (!node->left && !node->right) {
        int oddCounts = 0;
        for (int i = 1; i <= 9; i++) {
            if (count[i] % 2 != 0) oddCounts++;
        }
        if (oddCounts <= 1) ans = 1;  // Path is pseudo-palindromic
    } else {
        // Continue to child nodes
        ans += pseudoPalindromicPathsRecursive(node->left, count);
        ans += pseudoPalindromicPathsRecursive(node->right, count);
    }
    
    // Toggle the count back for this node's value
    count[node->val] ^= 1;
    
    return ans;
}

int pseudoPalindromicPaths (struct TreeNode* root) {
    int count[10] = {0};  // Count of node values from 1 to 9
    return pseudoPalindromicPathsRecursive(root, count);
}

File: 1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        def is_pseudo_palindromic(path: dict) -> bool:
            one_odd = False
            for count in path.values():
                if count % 2 == 1:
                    if one_odd:
                        return False
                    one_odd = True
            return True

        def helper(node: Optional[TreeNode], path: dict, num_pseudo_palindromic: list):
            if node:
                path[node.val] += 1
                if not node.left and not node.right:
                    if is_pseudo_palindromic(path):
                        num_pseudo_palindromic[0] += 1
                helper(node.left, path, num_pseudo_palindromic)
                helper(node.right, path, num_pseudo_palindromic)
                path[node.val] -= 1
                
        num_pseudo_palindromic = [0]
        helper(root, defaultdict(int), num_pseudo_palindromic)
        return num_pseudo_palindromic[0]